### ISSUE TITLE
[oneDNN v3.12.2] 3 E2E performance regressions vs v3.12 on BMG

### ISSUE THREAD

**Issue by mengfei25** [intel/torch-xpu-ops#4300]
## Description

After upgrading oneDNN from v3.12 to v3.12.2, 3 E2E inductor performance regressions were observed on BMG.
These have been verified with multiple independent re-runs (averaged results).

**Target:** oneDNN v3.12.2 (commit `d34e6c861d5dcbd1fa18904d0f80dcb0b6f74c6a`, workflow 28648466009)
**Baseline:** oneDNN v3.12 (commit `62a55dc7dc1735871a4ba2709f33bf8a7184fafd`, workflow 28579563092)
**Platform:** BMG Linux

## Summary

- **Inductor geomean ratio:** 1.0015x (no overall regression)
- **Verified drops:** 3 / 42 re-tested cases
- **Drop threshold:** inductor ratio (baseline_latency / target_latency) < 0.90

## Regressions

| Suite | Dtype | Mode | Model | Ratio |
|---|---|---|---|---|
| torchbench | float32 | inference | nanogpt | 0.8398 |
| huggingface | float16 | inference | openai/whisper-tiny | 0.8604 |
| huggingface | amp_bf16 | inference | openai/whisper-tiny | 0.8962 |

## Analysis

- **nanogpt (float32 inference):** ~16% slower inductor latency.
- **openai/whisper-tiny (float16 inference):** ~14% slower inductor latency.
- **openai/whisper-tiny (amp_bf16 inference):** ~10% slower inductor latency.

The whisper-tiny regressions in float16 and amp_bf16 may share a common root cause in oneDNN v3.12.2 affecting attention/conv operations for this model.

## Reproducer

```bash
# Setup pytorch with oneDNN v3.12.2
# can get prebuilt wheel from https://github.com/pytorch/pytorch/actions/runs/28574541829?pr=188787
# baseline v3.12 can use prebuilt pytorch release 2.13.0 `pip install torch==2.13.0 torchaudio torchvision --index-url https://download.pytorch.org/whl/test/xpu`
# cd pytorch

# nanogpt
python benchmarks/dynamo/torchbench.py -d xpu --performance --float32 --inference --only nanogpt --backend=inductor --cold-start-latency -n10 --disable-cudagraphs

# whisper-tiny float16
python benchmarks/dynamo/huggingface.py -d xpu --performance --float16 --inference --only openai/whisper-tiny --backend=inductor --cold-start-latency -n10 --disable-cudagraphs

# whisper-tiny amp_bf16
python benchmarks/dynamo/huggingface.py -d xpu --performance --amp --amp-dtype bfloat16 --inference --only openai/whisper-tiny --backend=inductor --cold-start-latency -n10 --disable-cudagraphs
```

## Environment

- Platform: BMG Linux
- PyTorch commit: d34e6c861d5dcbd1fa18904d0f80dcb0b6f74c6a (target) / 62a55dc7dc1735871a4ba2709f33bf8a7184fafd (baseline)
- Comparison method: inductor abs_latency, ratio = baseline/target, multiple runs averaged

**Comment by mengfei25** [comment]
## oneDNN Verbose Analysis

Collected `ONEDNN_VERBOSE=1` logs for the 3 regressed models and compared primitive-level execution times between v3.12 and v3.12.2 (iteration 6/10 → 7/10, averaged per shape).

### Methodology
1. Extract one iteration of verbose output (between benchmark iter 6/10 and 7/10)
2. Filter lines containing `primitive,exec,gpu` (actual kernel execution, excluding cache hits/creates)
3. Group by shape key (full oneDNN verbose descriptor minus the time), compute average time
4. Flag outliers where `v3.12 / v3.12.2 < 0.95` (i.e., v3.12.2 is slower)

### Results Summary

| Model | Outlier Shapes | Total Shapes | Top Contributor |
|-------|---------------|-------------|-----------------|
| whisper-tiny amp_bf16 | 15/27 | 27 | matmul `1x384:384x384` (9.5% of iter time, ratio=0.95) |
| whisper-tiny float16 | 16/19 | 19 | matmul `1x384:384x384` (20.4% of iter time, ratio=0.90) |
| nanogpt float32 | 0/11 | 11 | N/A (no primitive-level outliers) |

### Key Findings

**whisper-tiny (float16)** — 16/19 shapes regressed, dominated by small matmul shapes:
- `matmul 1x384:384x384` — ratio 0.90, 20.4% of iter time
- `matmul 1500x384:384x384` (with bias) — ratio 0.89, 11.2% of iter time
- `matmul 1x6x1x64:1x6x64x1500` (attention QK^T) — ratio 0.85, 5.7% of iter time
- `conv mb1_ic384oc384_iw3000ow1500kw3sw2` — ratio 0.85
- `conv mb1_ic80oc384_iw3000ow3000kw3sw1` — ratio 0.88

**whisper-tiny (amp_bf16)** — 15/27 shapes regressed, similar pattern:
- `matmul 1x384:384x384` — ratio 0.95, 9.5% of iter time
- `matmul 1x6x1x1500:1x6x1500x64` — ratio 0.81, 3.1%
- `matmul 1x6x1x64:1x6x64x1500` — ratio 0.88, 2.7%
- `softmax 1x6x1x1500` — ratio 0.94

**nanogpt (float32)** — No primitive-level regressions detected via verbose. The E2E regression may come from framework overhead or graph compilation differences rather than individual oneDNN primitives.

### benchdnn Reproducer

Generated benchdnn commands from outlier verbose lines using `verbose_converter.py`. **However, benchdnn with `--mode=P` cannot reproduce the performance drop for these shapes.** The individual primitive performance appears similar between v3.12 and v3.12.2 when tested in isolation via benchdnn.

This suggests the regression may be related to:
- Primitive creation/cache behavior differences in the full E2E execution context
- Memory allocation patterns or scratchpad management under load
- Interaction effects between multiple primitives sharing GPU resources
- Graph API partition overhead (whisper uses `sdp` graph partitions)

### Next Steps
- Investigate graph API compiled partition execution overhead
- Check if primitive cache hit/miss patterns differ between versions
- Profile GPU occupancy during full model inference to identify contention

**Comment by LuFinch** [comment]
All these three models use SDPA with host-scalar. 

Guilty commit :https://github.com/intel-innersource/libraries.performance.math.onednn/commit/77485d56b07f95f23765b434cf658f7cbf0bd512
